pkg/apis/ceph.rook.io/v1/types.go

```diff
@@ -964,9 +972,14 @@ type FilesystemSpec struct {
 	// +nullable
 	MetadataPool PoolSpec `json:"metadataPool"`
 
-	// The data pool settings
-	// +nullable
-	DataPools []PoolSpec `json:"dataPools"`
+	// (Deprecated in favor of namedDataPools) The data pool settings.
+	// +optional
+	DataPools []PoolSpec `json:"dataPools,omitempty"`
```

Comment:
Based on the discussion, I was thinking we would stick with only the `dataPools` and effectively be converting all of the FS data pools into named pools by changing this type to NamedPoolSpec. Since it's only internal json serialization by the operator, to the CRD it simply looks like the `name` is added, and no CRD type conversion needed.
```suggestion
	DataPools []NamedPoolSpec `json:"dataPools,omitempty"`
```

Reply:
In that case `name` parameter should be an optional for `NamedPoolSpec` - and if it is missed in spec we will use current default naming `<cephfsName>-data<index>`. So, then my question is still actual: do we need to create cephfs data pools with cephfs name prefix for `NamedPoolSpec`, e.g. `<cephfsName>-<nameFromSpec>` ?

Reply:
Also, such behaviour forces users to add pools with names add the end of current list **only**, otherwise it may brake current naming.

Reply:
Correct, the `name` would be optional, and if blank rook would set the generated pool name as today. 

If the set the `name` in a filesystem CR that already exists, and it is different than the existing generated pool name, it would certainly create a problem so we should add a check for that case. Otherwise, setting the pool name would cause a new pool to be created and added to the filesystem, and the default-named pool(s) would remain. 

> Also, such behaviour forces users to add pools with names add the end of current list only, otherwise it may brake current naming.

If they add the `name`, maybe we should enforce that the `name` be set on all the pools in the list. This would allow them to be reordered, even for the existing default-named pools.

Reply:
> If the set the name in a filesystem CR that already exists, and it is different than the existing generated pool name, it would certainly create a problem so we should add a check for that case. 

We can't add a check for that, since we are not tracking diff for spec (only showing it as simple diff string from comparer). Also, it is mostly same to situation, when user dropped old pool and added a new one named - real diff impossible to track. So, i'd rather to leave it up to user and warning them to avoid such situations, like add a different/new names for already present pools.

> If they add the name, maybe we should enforce that the name be set on all the pools in the list. This would allow them to be reordered, even for the existing default-named pools.

I dont think that it is needed. Since we already have an ordered list, user may just add new pools with predefined names and do not touch already present.

Again, my question is not answered: do we need to create cephfs data pools with cephfs name prefix for NamedPoolSpec, e.g. <cephfsName>-<nameFromSpec> ?

Reply:
> > If the set the name in a filesystem CR that already exists, and it is different than the existing generated pool name, it would certainly create a problem so we should add a check for that case.
> 
> We can't add a check for that, since we are not tracking diff for spec (only showing it as simple diff string from comparer). Also, it is mostly same to situation, when user dropped old pool and added a new one named - real diff impossible to track. So, i'd rather to leave it up to user and warning them to avoid such situations, like add a different/new names for already present pools.

Agreed, not sure how to check for that. 

> > If they add the name, maybe we should enforce that the name be set on all the pools in the list. This would allow them to be reordered, even for the existing default-named pools.
> 
> I dont think that it is needed. Since we already have an ordered list, user may just add new pools with predefined names and do not touch already present.

It's not really necessary, but it could help save some error scenarios or at least confusion around some pools having names and some not. We can skip the error check though, I'm fine with that. 

> Again, my question is not answered: do we need to create cephfs data pools with cephfs name prefix for NamedPoolSpec, e.g. - ?

This would certainly help ensure the pool names are unique when there are multiple filesystems with named pools, otherwise the pools could mistakenly overlap. My assumption with #9295 was that arbitrary pool names could be used, but sounds like you're interested in better names than data0, data1, ...

Btw I'm working on #9209 and the `NamedPoolSpec` will be useful for that implementation as well. I hope to have a PR open by tomorrow if not today. That change should set a foundation for what we want to do with this PR, so perhaps you wait for that to avoid too much implementation overlap.

Reply:
With #9363, hopefully this will be a much more straight forward implementation. The `NamedPoolSpec` is already passed down to the pool creation code, so it might basically look like:
- Change the types.go for dataPools to `NamedPoolSpec`
- In the filesystem.go, where you see `dataPoolNames[i]`, implement the name override if specified instead of using the generated name.

Reply:
> With #9363, hopefully this will be a much more straight forward implementation. The NamedPoolSpec is already passed down to the pool creation code, so it might basically look like:
Change the types.go for dataPools to NamedPoolSpec

Ok, let's wait. Then i'll just rebase, once your PR is merged.

> In the filesystem.go, where you see dataPoolNames[i], implement the name override if specified instead of using the generated name.

Actually, it is already done. But again, should we continue to use real pool names as `<cephfsName>-<poolNameFromSpec>` or just `<poolNameFromSpec>`?

Reply:
> But again, should we continue to use real pool names as `<cephfsName>-<poolNameFromSpec>` or just `<poolNameFromSpec>`?

I'm thinking `<cephfsName>-<poolNameFromSpec>` to prevent conflicts between filesystems. wdyt?

Reply:
Yeah, agree. I'll update.